tools/model_tools.R

```diff
@@ -39,3 +42,281 @@ interpolate_abundance <- function(abundance){
+  daily_forecasts$newmoonnumber <- newmoon_match_number[which_match]
+
+  newmoon_forecasts <- daily_forecasts %>% group_by(newmoonnumber) %>%
+         summarize(mintemp = min(mintemp, na.rm = T), maxtemp = max(maxtemp, na.rm = T), 
+                   meantemp = mean(meantemp, na.rm = T), 
+                   precipitation = sum(precipitation, na.rm = T))
+
+  return(newmoon_forecasts)
+}
+
+####################################################################################
+#' Build a weather forecast for model predictions by combining portal station data, 
+#' climate forecasts, and historic means, impose specified lag
+#' 
+#' @param start end year of data and beginning of forecast
+#' @param moons moon data
+#' @param lag newmoons by which weather data is lagged
+#' @param lead_time the number of newmoons into the future to obtain forecasts. 
+#' Max of 7. lag + lead_time should equal 12
+#' 
+#' @return a data.frame with 12 new moons of weather values
+#'
+fcast_weather = function(start = as.numeric(format(Sys.Date(), "%Y")), moons, lag = 6, lead_time = 6){
+  
+  newweather <- portalr::weather("newmoon",fill=TRUE) %>%
+    select(-c(locally_measured,battery_low)) %>% 
+    mutate(year = as.numeric(format(date, "%Y"))) %>%
+    filter(year>=start-5)
+  incompletes <- which(is.na(newweather$newmoonnumber))
+  if(length(incompletes) > 0 ){
+    newweather <- newweather[-incompletes, ]  
+  }
+
+  weatherforecast <- tail(newweather,lag) %>% select(-year, -date) %>%
+    bind_rows(get_climate_forecasts(lead_time = lead_time, moons = moons))
+
+  return(weatherforecast)
+}
+
+####################################################################################
+#' Fill in historic ndvi data to the complete timeseries being fit
+#'
+#' @details missing values during the time series are replaced using na.interp, 
+#'  missing values at the end of the time series are forecast using auto.arima with
+#'  seasonality (using Fourier transform)
+#'
+#' @param ndvi ndvi data
+#' @param moons moon data
+#' @param forecast_newmoons from the model metadata
+#' @param lag lag in new moons
+#'
+#' @return a data.frame with new moons and ndvi values
+#'
+fill_ndvi <- function(ndvi, moons, forecast_newmoons, lag){
+
+  min_hist_ndvi_nm <- min(ndvi$newmoonnumber)
+  max_hist_ndvi_nm <- max(ndvi$newmoonnumber)
+  hist_ndvi_nms <- min_hist_ndvi_nm:max_hist_ndvi_nm
+
+  hist_ndvi <- data.frame(newmoonnumber = hist_ndvi_nms, ndvi = NA)
+  nm_match <- match(hist_ndvi$newmoonnumber, ndvi$newmoonnumber)
+  hist_ndvi$ndvi <- ndvi$ndvi[nm_match]
+  ndvi_interp <- na.interp(hist_ndvi$ndvi)
+  hist_ndvi$ndvi <- as.numeric(ndvi_interp)
+
+  last_newmoon <- forecast_newmoons[1] - 1
+
+  if(max_hist_ndvi_nm < last_newmoon){
+
+    newmoons_to_fcast <- (max_hist_ndvi_nm + 1):last_newmoon
+    ndvi_fcast <- fcast_ndvi(hist_ndvi, moons, newmoons_to_fcast, lag, FALSE)
+    hist_ndvi <- rbind(hist_ndvi, ndvi_fcast)  
+  }
+  return(hist_ndvi)
+}
+
```

Comment:
Yeah yeah! Let's use this as our `fill` argument for ndvi over in portalr.

Reply:
just fyi, i'm going to need to generalize the function more, as right now it's based on newmoons only, and the ndvi can be managed as newmoons or months. it shouldn't be too much work, but it's more than just plopping the function over here.